### ISSUE TITLE
ClassVar[TypeAlias] in dataclass-like classes becomes Any

### ISSUE THREAD

**Issue by hynek** [python/mypy#17031]
**Bug Report**

When defining a type alias within a dataclass (or attrs) class, it has to be marked as `ClassVar`[^1] which causes the field to be treated as `Any`.

This bug has been already reported in https://github.com/python/mypy/issues/9238#issuecomment-1089065588 but never followed up.

**To Reproduce**

```python
from dataclasses import dataclass
from typing import ClassVar, TypeAlias


@dataclass
class Namespace:
    Alias: ClassVar[TypeAlias] = int
    
reveal_type(Namespace.Alias)
```

https://mypy-play.net/?mypy=latest&python=3.12&gist=e66e903f3b055fa8433707a5e8b99ebc

**Expected Behavior**

`main.py:16: note: Revealed type is "int"`

**Actual Behavior**

`main.py:16: note: Revealed type is "Any"`

**Your Environment**

See the playground. Python 3.12 on Mypy 1.9.0.

[^1]: The _attrs_ plugin crashes with an assertion error if this is forgotten, but that's a bug for another day.

**Comment by Hnasar** [comment]
I'm not sure the differences in semantics but `Alias: ClassVar[type[int]] = int` seems to be a workaround.

**Comment by richin13** [comment]
Also looks like pyright does get it right:

```
Type of "Namespace.Alias" is "type[int]"
```

**Comment by hynek** [comment]
Yes, it does. This forced me to switch my first project to Pyright. 😳

**Comment by tmke8** [comment]
> I'm not sure the differences in semantics but `Alias: ClassVar[type[int]] = int` seems to be a workaround.

I think the main difference is that `TypeAlias` can be used in type annotations but `type[...]` cannot: https://mypy-play.net/?mypy=latest&python=3.12&gist=43713abe4350c5ef288d2ff8f23e82ea

**Comment by tmke8** [comment]
This now works on my master:

```python
from dataclasses import dataclass
from typing_extensions import TypeAliasType

@dataclass
class A:
    T1 = TypeAliasType("T1", int)
    
x: A.T1 = 2
y: A.T1 = "foo"  # type error, as expected
```

Playground: https://mypy-play.net/?mypy=master&python=3.12&gist=1e50bf096b33af19e158d97dd828ae0c

It's based on `TypeAliasType` introduced in PEP 695.
